### ISSUE TITLE
Invalid Memory Address Error of Specific SQL Query

### ISSUE THREAD

**Issue by bajinsheng** [cockroachdb/cockroach#83973]
**Describe the problem**

An issue of invalid memory address.

**To Reproduce**

```sql
CREATE TABLE t3 (c0 DECIMAL);
SELECT t3.c0 FROM t3 WHERE (t3.rowid NOT BETWEEN (1) AND (1 + 1) AND (t3.rowid IS DISTINCT FROM -1000)) GROUP BY t3.c0; --ERROR: internal error: runtime error: invalid memory address or nil pointer dereference
```

**Expected behavior**
No error.

**Additional data / screenshots**

> ERROR: internal error: runtime error: invalid memory address or nil pointer dereference
SQLSTATE: XX000
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/util/errorutil/catch.go:29: ShouldCatch()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:227: func1()
GOROOT/src/runtime/panic.go:1038: gopanic()
GOROOT/src/runtime/panic.go:221: panicmem()
GOROOT/src/runtime/signal_unix.go:735: sigpanic()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/general_funcs.go:626: splitScanIntoUnionScansOrSelects()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/general_funcs.go:398: splitScanIntoUnionScans()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/groupby_funcs.go:341: SplitGroupByScanIntoUnionScans()
github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/opt/xform/explorer.og.go:3164: exploreDistinctOn()
github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/opt/xform/explorer.og.go:44: exploreGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.go:182: exploreGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:507: optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:247: Optimize()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:566: buildExecMemo()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:225: makeOptimizerPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1363: makeExecPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1045: dispatchToExecutionEngine()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:690: execStmtInOpenState()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:145: execStmt()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1892: func1()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1896: execCmd()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1818: run()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:757: ServeConn()
github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:724: func1()
GOROOT/src/runtime/asm_amd64.s:1581: goexit()
HINT: You have encountered an unexpected error.
Please check the public issue tracker to check whether this problem is
already tracked. If you cannot find it there, please report the error
with details by creating a new issue.
If you would rather not post publicly, please contact us directly
using the support form.
We appreciate your feedback.

**Environment:**
 - CockroachDB version [02dcb3816c4148ee15318a614901b266f1b8ae34]
 - Server OS: [Ubuntu 20.04]
 - Client app [`cockroach sql`]

**Additional context**

Jira issue: CRDB-17380

**Comment by mgartner** [comment]
Thanks for the report! This will be fixed by #82731 which should merge soon.

**Comment by bajinsheng** [comment]
Thanks for your prompt reply!
